File: 2025/python/06.py
import math
from dataclasses import dataclass
from pathlib import Path

import pytest

# NOTE: This input needs its dangling whitespace preserved!
test_input = """123 328  51 64 
 45 64  387 23 
  6 98  215 314
*   +   *   +  
"""


@dataclass
class Problem:
    numbers: list[int]
    operand: str

    def solve(self) -> int:
        match self.operand:
            case "+":
                result = sum(self.numbers)
            case "*":
                result = math.prod(self.numbers)
            case _:
                raise ValueError(f"Unknown operand: {self.operand}")
        return result


def part_one(input_data: str) -> int:
    lines = input_data.strip().splitlines()

    # Get operands and first line, then fill in the rest
    problems: list[Problem] = [
        Problem(
            numbers=[
                num,
            ],
            operand=op,
        )
        for (num, op) in zip(map(int, lines[0].split()), lines[-1].split())
    ]

    for line in lines[1:-1]:
        for i, num in enumerate(map(int, line.split())):
            problems[i].numbers.append(num)

    result = sum([p.solve() for p in problems])
    return result


def part_two(input_data: str) -> int:
    lines = input_data.splitlines()
    problems: list[Problem] = []

    # The operands mark the beginning x-coordinate of a new problem
    operand_indices = [i for i, c in enumerate(lines[-1]) if c in "+*"]

    # Split the input columns by operand indices
    for i in range(len(operand_indices)):
        start_idx = operand_indices[i]
        end_idx = (
            operand_indices[i + 1] - 1
            if i + 1 < len(operand_indices)
            else len(lines[0])
        )

        # Take columns, stack them and strip whitespace before converting to integers
        numbers = []
        for col_idx in range(start_idx, end_idx):
            column = "".join([line[col_idx] for line in lines[:-1]])
            numbers.append(int(column))

        operand = lines[-1][start_idx]
        problems.append(Problem(numbers=numbers, operand=operand))

    result = sum([p.solve() for p in problems])
    return result


@pytest.mark.parametrize("input_data, expected_output", [(test_input, 4_277_556)])
def test_part_one(input_data: str, expected_output: int):
    assert part_one(input_data) == expected_output


@pytest.mark.parametrize("input_data, expected_output", [(test_input, 3_263_827)])
def test_part_two(input_data: str, expected_output: int):
    assert part_two(input_data) == expected_output


if __name__ == "__main__":
    with Path("../06.in").open() as f:
        input_data = f.read()

    result1 = part_one(input_data)
    result2 = part_two(input_data)
    print(f"Part 1: {result1}")
    print(f"Part 2: {result2}")

File: 2025/cpp/src/06.cpp
#include <cassert>
import std;

using namespace std;

const string TEST_INPUT = "123 328  51 64 \n"
                          " 45 64  387 23 \n"
                          "  6 98  215 314\n"
                          "*   +   *   +  \n";

class Homework {
  public:
    enum class ReadDirection { RowWise, ColumnWise };
    Homework(const string& input_data) {
        auto stream = istringstream{input_data};
        string line;
        while (getline(stream, line)) {
            if (!line.empty()) {
                grid.push_back(line);
            }
        }

        // Identify operator indices indicating where columns are
        string& operators = grid.back();
        for (size_t col = 0; col < operators.size(); ++col) {
            if (operators[col] != ' ') {
                operator_indices.push_back(col);
            }
        }
    };

    auto solve(ReadDirection direction) const -> long long {
        vector<long long> results = {};
        // Split by operators first to create "Problems"
        for (size_t i = 0; i < operator_indices.size(); ++i) {
            size_t start_col = operator_indices[i];
            size_t end_col = (i + 1 < operator_indices.size())
                                 ? operator_indices[i + 1]
                                 : grid[0].size();

            vector<long long> numbers = {};

            if (direction == ReadDirection::ColumnWise) {
                for (size_t col_idx = start_col; col_idx < end_col; ++col_idx) {
                    string number_str = "";
                    for (size_t row_idx = 0; row_idx < grid.size() - 1;
                         ++row_idx) {
                        if (grid[row_idx][col_idx] != ' ') {
                            number_str += grid[row_idx][col_idx];
                        }
                    }
                    if (number_str.empty()) {
                        continue;
                    }
                    long long number = stoll(number_str);
                    numbers.push_back(number);
                }
            } else {
                for (size_t row_idx = 0; row_idx < grid.size() - 1; ++row_idx) {
                    string number_str = "";
                    for (size_t col_idx = start_col; col_idx < end_col;
                         ++col_idx) {
                        if (grid[row_idx][col_idx] != ' ') {
                            number_str += grid[row_idx][col_idx];
                        }
                    }
                    long long number = stoll(number_str);
                    numbers.push_back(number);
                }
            }

            char op = grid.back()[start_col];
            switch (op) {
            case '+': {
                long long sum =
                    accumulate(numbers.begin(), numbers.end(), 0LL, plus<>());
                results.push_back(sum);
                break;
            }
            case '*': {
                long long prod = accumulate(numbers.begin(), numbers.end(), 1LL,
                                            multiplies<>());
                results.push_back(prod);
                break;
            }
            }
        }
        long long sum = 0;
        for (auto res : results) {
            sum += res;
        }
        return sum;
    };

  private:
    vector<string> grid;
    vector<size_t> operator_indices;
};

auto part_one(const string& input_data) -> long long {
    auto homework = Homework(input_data);
    return homework.solve(Homework::ReadDirection::RowWise);
}

auto part_two(const string& input_data) -> long long {
    auto homework = Homework(input_data);
    return homework.solve(Homework::ReadDirection::ColumnWise);
}

auto test_part_one() -> void {
    auto expected = 4277556;
    auto result = part_one(TEST_INPUT);
    cout << "Got: " << result << ", Expected: " << expected << "\n";
    assert(result == expected);
}

auto test_part_two() -> void {
    auto expected = 3263827;
    auto result = part_two(TEST_INPUT);
    cout << "Got: " << result << ", Expected: " << expected << "\n";
    assert(result == expected);
}

auto main() -> int {
    test_part_one();
    test_part_two();

    string filepath = "../06.in";
    ifstream file(filepath);
    if (!file) {
        throw runtime_error("Failed to open file: " + filepath);
    }

    string in = string{istreambuf_iterator<char>(file), {}};

    long long result_one = part_one(in);
    long long result_two = part_two(in);
    cout << "Part one: " << result_one << "\n";
    cout << "Part two: " << result_two << "\n";
    return 0;
}
